Remove duplicated device feature detection function

iulianbarbu opened the pull request:
## Reason for This PR

Remove Duplicated Device-Feature detection function #1209

## Description of Changes

Methods **features** and **ack_features**, implemented in every virtio device that provide an implementation for VirtioDevice trait (e.g block, net, vsock), contain identical body, which is not device specific.

This PR removes code duplication, and move the aforementioned methods in the VirtioDevice trait, with an implementation for each method, which will be common to every virtio device.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] Either this PR is linked to an issue, or, the reason for this PR is
      clearly provided. 
- [x] The description of changes is clear and encompassing.
- [x] Either no docs need to be updated as part of this PR, or, the required
      doc changes are included in this PR. Docs in scope are all `*.md` files
      located either in the repository root, or in the `docs/` directory.
- [x] Either no code has been touched, or, code-level documentation for touched
      code is included in this PR.
- [x] Either no API changes are included in this PR, or, the API changes are
      reflected in `firecracker/swagger.yaml`.
- [x] Either the changes in this PR have no user impact, or, the changes in
      this PR have user impact and have been added to the `CHANGELOG.md` file.

sandreim commented on vmm/src/device_manager/mmio.rs:
```diff
@@ -367,6 +367,18 @@ mod tests {
+            0
+        }
+
+        fn set_acked_features(&mut self, acked_features: u64) {
```
I would do it this way, it's less code:
```suggestion
        fn set_acked_features(&mut self, _acked_features: u64) {}
```
or 
```suggestion
        fn set_acked_features(&mut self, _: u64) {}
```

sandreim commented on devices/src/virtio/mmio.rs:
```diff
@@ -41,12 +41,47 @@ pub trait VirtioDevice: Send {
+        self.set_acked_features(res);
+    }
+
+    // Means to expose inner state of a virtio device implementation to VirtioDevice trait
```
nit: I like to use full stop to end a comment.

sandreim commented on devices/src/virtio/mmio.rs:
```diff
@@ -41,12 +41,47 @@ pub trait VirtioDevice: Send {
+        }
+
+        let res = avail_features | v;
+        self.set_acked_features(res);
```
Less code suggestion:
```suggestion
        self.set_acked_features(avail_features | v);
```

acatangiu commented:
Also your PR checklist has nothing ticked on it 😄

iulianbarbu commented:
> Code changes look good, but you have to squash/rearrange your commits into something that makes sense.
> 
> Ideally, changes should be split logically in separate commits, while still building and not breaking functionality when used individually.
> 
> Off the top of my head you rearrange commits in this PR:
> 
> * two commits:
>   
>   1. add the generic implementation for the trait,
>   2. remove the specific so it uses the generic.
> 
> or
> 
> * squash everything in a single commit since the changes are not extensive and can be easily followed.

iulianbarbu commented:
Squashed into a single commit. 
Ticked the items of the checklist. 
Also, added some documentation comments for the newly introduced methods from VirtioDevice trait, as @tamionv suggested.

acatangiu commented:
Please also update the commit message to capture all the changes, right now the information from the squashed commit messages is lost.

dianpopa commented on devices/src/virtio/mmio.rs:
```diff
@@ -449,6 +489,19 @@ mod tests {
+
+        #[allow(clippy::cast_lossless)]
+        fn get_acked_features(&self) -> u64 {
+            self.acked_features as u64
```
Why not `u64::from(self.acked_features)` in order to avoid the `#[allow(clippy::cast_lossless)]`?

iulianbarbu replied:
Yep, it's better with your suggestion. Updated.

dianpopa commented on devices/src/virtio/mmio.rs:
```diff
@@ -33,20 +33,60 @@ const MMIO_VERSION: u32 = 2;
 
     /// Acknowledges that this set of features should be enabled.
-    fn ack_features(&mut self, page: u32, value: u32);
+    fn paged_acked_features(&mut self, page: u32, value: u32) {
```
I do not like the `paged` suffix from the functions name. The page argument dictates which bits should be taken into account (high or low). So it plays the role of a select.
I think the names of the functions should go like this:
`fn avail_features` -> gets available features, this is a getter so it should have the name of the field it gets
`fn acked_features` -> Ok
`fn select_acked_features` -> function for selecting acknowledged features using a select value
`fn select_avail_features` -> get available features using a select value

acatangiu replied:
`select_*` sounds like the action of selecting `*`.

The virtio documentation refers the upper and lower 32bit parts of features as `pages`. How about:

`fn avail_features() -> u64` - gets `self.avail_features`
`fn set_avail_features(val: u64)` - sets `self.avail_features`
`fn acked_features() -> u64` - gets already acked features (`self.acked_features`)
`fn set_acked_features(val: u64)` - sets already acked features (`self.acked_features`)

`fn ack_features_by_page(page: u32, val: u32)` - the action of acknowledging the features in `val` from page `page`
`fn avail_features_by_page(page:u32)` - gets avail features in page `page`

?

Notice the difference in the ACK case, getter and setter work on **acked** values, while the `fn ack_features_by_page()` has **ack** as a verb, the function is doing the action of acknowledging.

iulianbarbu replied:
I am inclined to @acatangiu suggestion. Also, the `fn ack_features_by_page()` method suggests verbatim the action which it should be responsible of, information which was lost once I've changed the original method name.

dianpopa replied:
I am ok with that proposal too.

iulianbarbu replied:
Updated as discussed.

acatangiu commented on devices/src/virtio/mmio.rs:
```diff
@@ -33,20 +33,60 @@ const MMIO_VERSION: u32 = 2;
+            v &= !unrequested_features;
+        }
+
+        self.set_acked_features(avail_features | v);
```
Good catch @dianpopa !

```suggestion
        self.set_acked_features(self.acked_features() | v);
```
@iulianbarbu please also add a test that fails when features are not correctly ACKed.
We should've had a test that failed with this version of the PR.

iulianbarbu replied:
Fixed the bug and added a test for it.

acatangiu commented on devices/src/virtio/mmio.rs:
```diff
@@ -338,7 +380,12 @@ impl BusDevice for MmioDevice {
-                            self.device.ack_features(self.acked_features_select, v);
+                            // First, set the device available features as those acknowledged by driver.
+                            // This should result in ack_features == self.acked_features_select + v.
+                            self.device
```
This is incorrect.
`self.acked_features_select` hold the value for selecting either the upper or lower 32bit dword of the `acked_features`.

Don't change `avail_features` at this step of the virtio config.

acatangiu commented on devices/src/virtio/mmio.rs:
```diff
@@ -407,6 +454,7 @@ mod tests {
 
     struct DummyDevice {
         acked_features: u32,
+        avail_features: u32,
```
You can make this an `u64` to be easier to test with the proper handlers. It was an `u32` to hack a simple test which is no longer required.

dianpopa commented:
@acatangiu can you please review your requested changes?

dianpopa commented on devices/src/virtio/mmio.rs:
```diff
@@ -1,4 +1,3 @@
-// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Why is this removed?

iulianbarbu replied:
Because of force pushing in hurry. Sorry.
Please re-review.

acatangiu commented:
@iulianbarbu
> Because of force pushing in hurry. Sorry.
Please re-review.

So you fixed it by force pushing in a hurry again without pulling the new upstream :laughing:
